### PR TITLE
CoreAudio_exclusive: fix #4478, #3072, #5238 and other improvements

### LINKED ISSUE #4478: Coreaudio_exclusive only work with audio-format=s16 on a 24bit external DAC
### mpv version and platform

MacOS 10.12.5 (16F73), MPV Git 7e889e5

### Reproduction steps
Run with `-no-config -audio-format=s16 -audio-exclusive=yes` and `-no-config -audio-format=s24 -audio-exclusive=yes`

The specification of my DAC. It is 24 bit compatible.
"[   0.292][v][ao/coreaudio_exclusive] -  192000.0Hz 24bit mcpl [4][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  176400.0Hz 24bit mcpl [4][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  96000.0Hz 24bit mcpl [4][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  88200.0Hz 24bit mcpl [4][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  48000.0Hz 24bit mcpl [4][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  44100.0Hz 24bit mcpl [4][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  192000.0Hz 16bit mcpl [12][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  176400.0Hz 16bit mcpl [12][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  96000.0Hz 16bit mcpl [12][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  88200.0Hz 16bit mcpl [12][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  48000.0Hz 16bit mcpl [12][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  44100.0Hz 16bit mcpl [12][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  192000.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  176400.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  96000.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  88200.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  48000.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  44100.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   0.292][v][ao/coreaudio_exclusive] -  192000.0Hz 16bit mcpl [76][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  176400.0Hz 16bit mcpl [76][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  96000.0Hz 16bit mcpl [76][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  88200.0Hz 16bit mcpl [76][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  48000.0Hz 16bit mcpl [76][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)
[   0.292][v][ao/coreaudio_exclusive] -  44100.0Hz 16bit mcpl [76][4bpp][1fbp][4bpf][2ch] int LE S packed (s16)"

### Expected behavior
Both `audio-format=s16` and `audio-format=s24`should work properly
### Actual behavior
`audio-format=s16` works fine. `audio-format=s24` terminates. 

Error message:
"[   1.009][v][ao/coreaudio_exclusive] virtual format 96000.0Hz 24bit mcpl [68][8bpp][1fbp][8bpf][2ch] int LE S (-)
[   1.009][e][ao/coreaudio_exclusive] hardware format not supported
[   1.009][e][ao] Failed to initialize audio driver 'coreaudio'
[   1.009][e][cplayer] Could not open/initialize audio device -> no sound.
[   1.009][v][ad] Uninit audio decoder.
[   1.009][v][af] Removing filter lavrresample 
[   1.010][i][cplayer] Audio: no audio
[   1.010][v][cplayer] playback restart complete
[   1.010][v][cplayer] EOF code: 1  
[   1.010][v][vd] Uninit video.
[   1.010][v][cplayer] finished playback, audio output initialization failed (reason 4)"

### Log file

[s16.log](http://sprunge.us/FfNV)
[s24.log](http://sprunge.us/RUdJ)

### Sample files

Sample files needed to reproduce this issue can be uploaded to https://0x0.st/
or similar sites. (Only needed if the issue cannot be reproduced without it.)

### PR DESCRIPTION
**Pull request by macdavis**
I am not a programmer, and only with very limited coding knowledge. If you have any suggestion, please kindly tell me exactly what I should do. Otherwise, I probably can't understand what you are talking about. Sorry.

There is one minor issue I'm not sure how to solve. I tried to modify `mp_tag_str_buf` to show hexadecimal number, but it prints "0x30x1" rather than "0x301".

I acknowledge that I received great help from the following parties:
	@downthomas,
	@wm4,
	@badgerkin;

and I took reference from the following code.
	https://github.com/cmus/cmus/blob/master/op/coreaudio.c (cmus, GNU General Public License v2.0)

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

**Comment by Akemi** [comment]
since you mentioned you took references from a GPL licensed code we can't take your changes like that. first of all, it's not clear what you mean with reference. in the worst case we would violate the license.

i am not an expert on licenses and can't say 'how much reference' taking would even be okay.

i would also prefer to get some more incremental PRs, if possible. fixing or adding one thing after another.

there are also some othe rissued, like mod changed files or removed file ends.

**Comment by jeeb** [comment]
@macdavis sorry for there being no replies for a while, but there's one thing that generally keeps people from looking at the code in this PR.

If the code is based on code of a GPL project, that cannot become LGPL without the original project consenting. Currently the coreaudio audio output module is licensed under the LGPLv2.1+ license, and the GPL license is a more limited license than LGPLv2.1+. Thus, if you actually took code from that other project - unless we make this module only available under the GPL only - we cannot take it in.

Thus I would like an explanation of what the "took reference" meant.

Otherwise, thank you for doing an explanation of the logical changes, though. That is one thing that can be read without worries.

**Comment by macdavis** [comment]
This is the original code that I took reference from,

```
static OSStatus coreaudio_set_buffer_size(AudioUnit au, AudioStreamBasicDescription desc, int *frame_size)
{
	AudioValueRange value_range = {0, 0};
	UInt32 property_size = sizeof(AudioValueRange);
	OSStatus err = AudioUnitGetProperty(au,
					    kAudioDevicePropertyBufferFrameSizeRange,
					    kAudioUnitScope_Global,
					    0,
					    &value_range,
					    &property_size);
	if (err != noErr)
		return err;
`

	UInt32 buffer_frame_size = value_range.mMaximum;
	err = AudioUnitSetProperty(au,
				   kAudioDevicePropertyBufferFrameSize,
				   kAudioUnitScope_Global,
				   0,
				   &buffer_frame_size,
				   sizeof(buffer_frame_size));
	if (err != noErr)
		d_print("Failed to set maximum buffer size: %d\n", err);

	property_size = sizeof(buffer_frame_size);
	err = AudioUnitGetProperty(au,
				   kAudioDevicePropertyBufferFrameSize,
				   kAudioUnitScope_Global,
				   0,
				   &buffer_frame_size,
				   &property_size);
	if (err != noErr) {
		d_print("Cannot get the buffer frame size: %d\n", err);
		return err;
	}

	buffer_frame_size *= desc.mBytesPerFrame;

	// We set the frame size to a power of two integer that
	// is larger than buffer_frame_size.
	while (*frame_size < buffer_frame_size + 1) {
		*frame_size <<= 1;
	}

	return noErr;
}
```


This is what I wrote,
```
OSStatus ca_get_frame_buffer_size(struct ao *ao, AudioDeviceID device)
 {
     AudioValueRange value_range = {0, 0};
     OSStatus err = CA_GET_O(device, kAudioDevicePropertyBufferFrameSizeRange, &value_range);
     MP_VERBOSE(ao, "Frame buffer size: %g to %g frames\n", value_range.mMinimum, value_range.mMaximum);
     return err;
 }
```

and

```
OSStatus ca_set_frame_buffer_size(struct ao *ao, AudioDeviceID device, int *buffersize)
 {
     // Reference: https://github.com/cmus/cmus/blob/master/op/coreaudio.c
     AudioValueRange value_range = {0, 0};
     OSStatus err1 = CA_GET_O(device, kAudioDevicePropertyBufferFrameSizeRange, &value_range);
     MP_VERBOSE(ao, "Frame buffer size: %g to %g frames\n", value_range.mMinimum, value_range.mMaximum);

      OSStatus err = CA_SET(device, kAudioDevicePropertyBufferFrameSize, buffersize);

      // e.g. 16 bit SPDIF AC-3 has a static buffer size of 1536 frames.

    if (value_range.mMinimum == value_range.mMaximum){
         *buffersize = value_range.mMinimum;
         MP_VERBOSE(ao, "Set frame buffer size to %d frames\n", *buffersize);
     }else{
         if ((*buffersize >= value_range.mMinimum) && (*buffersize <= value_range.mMaximum)){
             MP_VERBOSE(ao, "Set frame buffer size to %d frames\n", *buffersize);
         }else if (*buffersize < value_range.mMinimum){
             *buffersize = value_range.mMinimum;
             MP_VERBOSE(ao, "Target frame size it invalid, set to %d (min) frames\n", *buffersize);
         }else{
             *buffersize = value_range.mMaximum;
             MP_VERBOSE(ao, "Target frame size it invalid, set to %d (max) frames\n", *buffersize);
         }
     }

      return err & err1;
 }
```

**Comment by wm4** [comment]
Yeah, doesn't look like it's copied code. So it should be OK as far as Copyright goes. Maybe still mention the URL to the other project in the commit message and how it was looked at for reference.

File: audio/out/ao_coreaudio_exclusive.c [hunk 1]
```diff
@@ -75,13 +75,76 @@ struct priv {
...
+    }else if ((integer_mode_avaliable == 0) && (maxbitdepth_physical == 32)){
+        device_type = 4; // 32 Bit device w/o Integer Mode.
+    }else if ((integer_mode_avaliable == 0) && (maxbitdepth_physical == 24)){
+        device_type = 10; // 24 Bit device w/o Integer Mode.
+    }else if ((integer_mode_avaliable == 0) && (maxbitdepth_physical == 20)){
+        device_type = 11; // 20 Bit device w/o Integer Mode.
+    }else if ((integer_mode_avaliable == 1) && (maxbitdepth_physical == 20)
+        && (max_mBytesPerPacket == 8) && (aligned_high == 0)){
+        device_type = 12; // 20 bit aligned low devce (shown in optical output).
+    }
```
Comment (wm4) [inline audio/out/ao_coreaudio_exclusive.c]: I don't think there should be such a device_type integer. They're magical values (and a _lot_ of them), what these magical values mean is widely spread over the entire source, and many of them seem to be used only in 1 place.

It looks like just keeping around integer_mode_avaliable and maxbitdepth_physical as fields would already be a large improvement.

Comment (wm4) [reply audio/out/ao_coreaudio_exclusive.c]: Looks like integer_mode_avaliable has a typo, btw.

Comment (macdavis) [reply audio/out/ao_coreaudio_exclusive.c]: device_type is needed in render_cb_compressed, find_best_format and init, but I agree, we don't have to consider so many different device types at this stage. After all, there are only few 20 Bit or aligned high devices. We could add support later, if anyone raises this issue.

Comment (wm4) [reply audio/out/ao_coreaudio_exclusive.c]: What I mean is that you shouldn't enumerate all these cases explicitly. Instead, use the capabilities of the device wherever it matters.

Comment (macdavis) [reply audio/out/ao_coreaudio_exclusive.c]: I tried, but each time, it needs to inquire`kAudioStreamPropertyAvailablePhysicalFormats` and do the for loop, which add 1-2 sec delay on my computer.

Comment (wm4) [reply audio/out/ao_coreaudio_exclusive.c]: Oh come on. Cache it. This is not hard.

Comment (macdavis) [reply audio/out/ao_coreaudio_exclusive.c]: Sorry, I don't understand what you mean. As I said, I am not a programmer, and this is already the best I can do.

Comment (wm4) [reply audio/out/ao_coreaudio_exclusive.c]: This is exactly what nobody ever wants to hear in a code review.

**Comment by Akemi** [comment]
@ruihe774 wondering if you are interested in working on this or rather sort all the things out?

**Comment by ruihe774** [comment]
> @ruihe774 wondering if you are interested in working on this or rather sort all the things out?

I may have a look at it at some time, though I personally don't think "bit exactness" really matters.

**Pull request closed without merging**